### PR TITLE
Add minimum commission rate to x/staking

### PR DESCRIPTION
**Pull request by sunnya97**
## Description

closes: https://github.com/osmosis-labs/osmosis/issues/114

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: x/staking/keeper/params.go [hunk 1]
```diff
@@ -39,6 +39,12 @@ func (k Keeper) BondDenom(ctx sdk.Context) (res string) {
 	return
 }
 
+// MinCommissionRate - Minimum validator commission rate
+func (k Keeper) MinCommissionRate(ctx sdk.Context) (res sdk.Dec) {
+	k.paramstore.Get(ctx, types.KeyMinCommissionRate, &res)
+	return
+}
+
 // Get all parameteras as types.Params
```
Comment (sunnya97) [inline x/staking/keeper/params.go:48]:
```suggestion
// Get all parameters as types.Params
```

File: x/staking/keeper/validator_test.go [hunk 2]
```diff
@@ -1083,13 +1089,15 @@ func TestUpdateValidatorCommission(t *testing.T) {
...
 		{val2, sdk.NewDecWithPrec(3, 1), true},
+		{val2, sdk.NewDecWithPrec(1, 2), true},
 		{val2, sdk.NewDecWithPrec(2, 1), false},
 	}
 
 	for i, tc := range testCases {
 		commission, err := app.StakingKeeper.UpdateValidatorCommission(ctx, tc.validator, tc.newRate)
 
 		if tc.expectedErr {
+			fmt.Println(app.StakingKeeper.MinCommissionRate(ctx))
```
Comment (sunnya97) [inline x/staking/keeper/validator_test.go:1100]:
```suggestion
```

File: x/staking/keeper/validator_test.go [hunk 3]
```diff
@@ -1059,6 +1059,12 @@ func TestUpdateValidatorCommission(t *testing.T) {
 	app, ctx, _, addrVals := bootstrapValidatorTest(t, 1000, 20)
 	ctx = ctx.WithBlockHeader(tmproto.Header{Time: time.Now().UTC()})
 
+	params := app.StakingKeeper.GetParams(ctx)
+
+	params.MinCommissionRate = sdk.NewDecWithPrec(5, 2)
```
Comment (ValarDragon) [inline x/staking/keeper/validator_test.go:1064]: Can you generate this from a string, pretty hard to grok this

File: x/staking/keeper/validator_test.go [hunk 4]
```diff
@@ -1083,6 +1089,7 @@ func TestUpdateValidatorCommission(t *testing.T) {
 		{val2, sdk.NewDecWithPrec(-1, 1), true},
 		{val2, sdk.NewDecWithPrec(4, 1), true},
 		{val2, sdk.NewDecWithPrec(3, 1), true},
+		{val2, sdk.NewDecWithPrec(1, 2), true},
```
Comment (ValarDragon) [inline x/staking/keeper/validator_test.go:1092]: Can a comment be added after this test case for why its failing?

File: x/staking/types/params.go [hunk 5]
```diff
@@ -191,3 +204,19 @@ func validateBondDenom(i interface{}) error {
...
+	v, ok := i.(sdk.Dec)
+	if !ok {
+		return fmt.Errorf("invalid parameter type: %T", i)
+	}
+
+	if v.IsNegative() {
+		return fmt.Errorf("minimum commission rate cannot be negative: %s", v)
+	}
+	if v.GT(sdk.OneDec()) {
+		return fmt.Errorf("minimum commission rate cannot be greater than 100%%")
```
Comment (ValarDragon) [inline x/staking/types/params.go:218]:
```suggestion
		return fmt.Errorf("minimum commission rate cannot be greater than 100%")
```

Comment (sunnya97) [reply x/staking/types/params.go:218]: I think it has to be like this

https://programming.guide/go/fmt-printf-percent-sign.html

Comment (ValarDragon) [reply x/staking/types/params.go:218]: oh im surprised fmt.Errorf doesnt fail lint if theres no extra args

File: x/staking/keeper/validator_test.go (hunk 3 above)
Comment (sunnya97) [reply x/staking/keeper/validator_test.go:1064]:
```suggestion
	params.MinCommissionRate = sdk.MustNewDecFromStr("0.05")
```

**Comment by sunnya97** [comment]
@ValarDragon is there a way to fix the protobuf/breakage check to test against the branch being merged into instead of master?

**Comment by sunnya97** [comment]
Fixed the protobuf breakage test to test against `osmo-v0.42.x` branch.

**Pull request merged**